session/nontransactional.go

```diff
@@ -0,0 +1,509 @@
+	}
+	if rs != nil {
+		rs.Close()
+	}
+	return ""
+}
+
+func buildShardJobs(ctx context.Context, stmt *ast.NonTransactionalDeleteStmt, se Session,
+	selectSQL string, shardColumn *table.Column) ([]job, error) {
+	logutil.Logger(ctx).Info("Non-transactional delete, select SQL", zap.String("selectSQL", selectSQL))
+	var shardColumnCollate string
+	if shardColumn != nil {
+		shardColumnCollate = shardColumn.Collate
+	} else {
+		shardColumnCollate = ""
+	}
+
+	rss, err := se.Execute(context.TODO(), selectSQL)
+	if err != nil {
+		return nil, err
+	}
+	if len(rss) > 1 {
+		return nil, errors.New("Non-transactional delete, more than 1 record sets")
+	}
+	rs := rss[0]
+	defer rs.Close()
+
+	batchSize := int(stmt.Limit)
+	if batchSize <= 0 {
+		return nil, errors.New("Non-transactional delete, batch size should be positive")
+	}
+	jobCount := 0
+	jobs := make([]job, 0)
+	currentSize := 0
+	var currentStart, currentEnd types.Datum
+
+	chk := rs.NewChunk(nil)
+	for {
+		err = rs.Next(context.TODO(), chk)
+		if err != nil {
+			return nil, err
+		}
+
+		// last chunk
+		if chk.NumRows() == 0 {
+			if currentSize > 0 {
+				// there's remaining work
+				jobs = append(jobs, job{jobID: jobCount, start: currentStart, end: currentEnd, jobSize: currentSize})
+			}
+			break
+		}
+
+		newStart := chk.GetRow(0).GetDatum(0, &rs.Fields()[0].Column.FieldType)
+
+		// end last batch if: (1) current start != last end (2) current size >= batch size
+		if currentSize >= batchSize {
+			cmp, err := newStart.Compare(se.GetSessionVars().StmtCtx, &currentEnd, collate.GetCollator(shardColumnCollate))
+			if err != nil {
+				return nil, err
+			}
+			if cmp != 0 {
+				jobs = append(jobs, job{jobID: jobCount, start: currentStart, end: currentEnd, jobSize: currentSize})
+				jobCount++
+				currentSize = 0
+			}
+		}
+
+		// a new batch
+		if currentSize == 0 {
+			currentStart = *newStart.Clone()
+		}
+
+		currentSize += chk.NumRows()
+		currentEndPointer := chk.GetRow(chk.NumRows()-1).GetDatum(0, &rs.Fields()[0].Column.FieldType)
+		currentEnd = *currentEndPointer.Clone()
+	}
+
+	return jobs, nil
+}
+
+func buildSelectSQL(stmt *ast.NonTransactionalDeleteStmt, se Session) (*ast.TableName, string, *table.Column, error) {
+	// only use the first table
+	// TODO: return error if there are multiple tables
+	if stmt.DeleteStmt.TableRefs == nil || stmt.DeleteStmt.TableRefs.TableRefs == nil {
+		return nil, "", nil, errors.New("table reference is nil")
+	}
+	if stmt.DeleteStmt.TableRefs.TableRefs.Right != nil {
+		return nil, "", nil, errors.New("Non-transactional delete doesn't support multiple tables")
+	}
+	tableSource, ok := stmt.DeleteStmt.TableRefs.TableRefs.Left.(*ast.TableSource)
+	if !ok {
+		return nil, "", nil, errors.New("Non-transactional delete, table source not found")
+	}
+	tableName, ok := tableSource.Source.(*ast.TableName)
+	if !ok {
+		return nil, "", nil, errors.New("Non-transactional delete, table name not found")
```

Comment:
How to construct these sqls that produce errors like "Non-transactional delete, table name not found" and above.

Reply:
IDK. I hope we will never see this error :)